File: subprojects/sc/android/app/src/main/java/com/redcarrottt/sc/internal/ControlMessageReceiver.java
package com.redcarrottt.sc.internal;

import android.util.Log;

import com.redcarrottt.testapp.Logger;

import java.util.ArrayList;

/* Copyright (c) 2018, contributors. All rights reserved.
 *
 * Contributor: Gyeonghwan Hong <redcarrottt@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class ControlMessageReceiver {
    private static final String kTag = "ControlMessageReceiver";
    private static final String kThreadName = "Control Message Receiving";

    public void startReceivingThread() {
        this.mReceivingThread = new ReceivingThread();
        this.mReceivingThread.start();
    }

    public void stopReceivingThread() {
        this.mReceivingThread.finish();
    }

    // Thread
    private class ReceivingThread extends Thread {
        @Override
        public void run() {
            this.setName("ControlMsgReceiver");
            this.mIsOn = true;
            Logger.THREAD_LAUNCH(kThreadName);

            while (this.mIsOn) {
                this.loopInternal();
            }

            Logger.THREAD_FINISH(kThreadName);
            this.mIsOn = false;
        }

        public boolean loopInternal() {
            byte[] messageBuffer = new byte[10 * 1024];
            int res = Core.singleton().receive(messageBuffer, true);
            if (res <= 0) {
                return false;
            }

            String message = new String(messageBuffer).trim();

            // Find separator location (between first line and other lines)
            Log.d(kTag, "Control message incoming: " + message);
            int separatorPos = message.indexOf('\n');

            // Divide the message into first line & other lines
            String firstLine = message.substring(0, separatorPos);
            String otherLines = message.substring(separatorPos + 1);

            int controlMessageCode = Integer.parseInt(firstLine);

            switch (controlMessageCode) {
                case ControlMessageProtocol.CMCode.kConnect:
                case ControlMessageProtocol.CMCode.kSleep:
                case ControlMessageProtocol.CMCode.kWakeup:
                case ControlMessageProtocol.CMCode.kDisconnectAck: {
                    // Normal type
                    int adapterId = Integer.parseInt(otherLines);
                    onReceiveNormalMessage(controlMessageCode, adapterId);
                    break;
                }
                case ControlMessageProtocol.CMCode.kDisconnect: {
                    // Disconnect type
                    // Divide the message into second line, third line, fourth line
                    separatorPos = otherLines.indexOf('\n');
                    String secondLine = otherLines.substring(0, separatorPos);
                    String thirdFourthLine = otherLines.substring(separatorPos + 1);
                    separatorPos = thirdFourthLine.indexOf('\n');
                    String thirdLine = thirdFourthLine.substring(0, separatorPos);
                    String fourthLine = thirdFourthLine.substring(separatorPos + 1);

                    int adapterId = Integer.parseInt(secondLine);
                    int finalSeqNoControl = Integer.parseInt(thirdLine);
                    int finalSeqNoData = Integer.parseInt(fourthLine);

                    onReceiveDisconnectMessage(adapterId, finalSeqNoControl, finalSeqNoData);
                    break;
                }
                case ControlMessageProtocol.CMCode.kPriv: {
                    // Priv type
                    onReceivePrivateMessage(otherLines);
                    break;
                }
                default: {
                    Logger.ERR(kTag, "Unknown control message code (" + controlMessageCode + ")"
                            + "!\n" + message);
                    break;
                }
            }

            return true;
        }

        public void finish() {
            this.mIsOn = false;
        }

        // Constructor
        public ReceivingThread() {
            this.mIsOn = false;
        }

        // Attributes
        private boolean mIsOn;
    }

    private void onReceiveNormalMessage(int controlMessageCode, int adapterId) {
        switch (controlMessageCode) {
            case ControlMessageProtocol.CMCode.kConnect: {
                Logger.VERB(kTag, "Receive(Control Msg): Request(Connect " + adapterId + ")");
                NetworkSwitcher.singleton().connectAdapterByPeer(adapterId);
                break;
            }
            case ControlMessageProtocol.CMCode.kSleep: {
                Logger.VERB(kTag, "Receive(Control Msg): Request(Sleep " + adapterId + ")");
                NetworkSwitcher.singleton().sleepAdapterByPeer(adapterId);
                break;
            }
            case ControlMessageProtocol.CMCode.kWakeup: {
                Logger.VERB(kTag, "Receive(Control Msg): Request(WakeUp " + adapterId + ")");
                NetworkSwitcher.singleton().wakeUpAdapterByPeer(adapterId);
                break;
            }
            case ControlMessageProtocol.CMCode.kDisconnectAck: {
                Logger.VERB(kTag, "Receive(Control Msg): Request(DisconenctAck " + adapterId + ")");
                ClientAdapter disconnectAdapter = Core.singleton().findAdapterById(adapterId);
                if (disconnectAdapter == null) {
                    Logger.WARN(kTag, "Cannot find adapter " + adapterId);
                } else {
                    disconnectAdapter.peerKnowsDisconnectingOnPurpose();
                }
                break;
            }
        }
    }

    private void onReceiveDisconnectMessage(int adapterId, int finalSeqNoControl, int
            finalSeqNoData) {
        Logger.VERB(kTag, "Receive(Control Msg): Request(Disconnect " + adapterId + " / " +
                "final_seq_no_control=" + finalSeqNoControl + " / final_seq_no_data=" +
                finalSeqNoData + ")");
        NetworkSwitcher.singleton().disconnectAdapterByPeer(adapterId, finalSeqNoControl,
                finalSeqNoData);
    }

    private void onReceivePrivateMessage(String contents) {
        // Find separator location
        int separatorPos = contents.indexOf('\n');

        // Divide the message into second line & other lines
        String secondLine = contents.substring(0, separatorPos);
        String privateMessage = contents.substring(separatorPos + 1);

        int privateType = Integer.parseInt(secondLine);

        // Notify the private message
        for (ControlMessageListener listener : mControlMessageListeners) {
            listener.onReceiveControlMessage(privateType, privateMessage);
        }
    }

    public void addControlMessageListener(ControlMessageListener listener) {
        this.mControlMessageListeners.add(listener);
    }

    public ControlMessageReceiver() {
        this.mReceivingThread = null;
        this.mControlMessageListeners = new ArrayList<>();
    }

    // Attributes
    private ReceivingThread mReceivingThread;
    private ArrayList<ControlMessageListener> mControlMessageListeners;
}

File: subprojects/linux/core/src/ControlMessageReceiver.cpp
/* Copyright 2017-2018 All Rights Reserved.
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *
 * [Contact]
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0(the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../inc/ControlMessageReceiver.h"

#include "../inc/Core.h"
#include "../inc/NetworkSwitcher.h"

#include "../../common/inc/DebugLog.h"

#include <iostream>
#include <string.h>
#include <string>
#include <sys/types.h>
#include <thread>
#include <unistd.h>

#define THREAD_NAME "Control Message Receiving"

using namespace sc;

void ControlMessageReceiver::start_receiving_thread(void) {
  this->mReceivingThread = new std::thread(
      std::bind(&ControlMessageReceiver::receiving_thread_loop, this));
  this->mReceivingThread->detach();
}

void ControlMessageReceiver::stop_receiving_thread(void) {
  this->mReceivingThreadOn = false;
}

void ControlMessageReceiver::receiving_thread_loop(void) {
  this->mReceivingThreadOn = true;
  LOG_THREAD_LAUNCH(THREAD_NAME);

  while (this->mReceivingThreadOn) {
    this->receiving_thread_loop_internal();
  }

  LOG_THREAD_FINISH(THREAD_NAME);
  this->mReceivingThreadOn = false;
}

bool ControlMessageReceiver::receiving_thread_loop_internal(void) {
  void *message_buffer = NULL;
  Core::singleton()->receive(&message_buffer, true);
  if (message_buffer == NULL) {
    return false;
  }

  std::string message((char *)message_buffer);

  // Find separator location (between first line and other lines)
  std::size_t separator_pos = message.find('\n');

  // Divide the message into first line & other lines
  std::string first_line(message.substr(0, separator_pos));
  std::string other_lines(message.substr(separator_pos));

  int control_message_code = std::stoi(first_line);

  switch (control_message_code) {
  case CMCode::kCMCodeConnect:
  case CMCode::kCMCodeSleep:
  case CMCode::kCMCodeWakeUp:
  case CMCode::kCMCodeDisconnectAck: {
    // Normal type
    int adapter_id = std::stoi(other_lines);
    this->on_receive_normal_message(control_message_code, adapter_id);
    break;
  }
  case CMCode::kCMCodeDisconnect: {
    // Disconnect type
    // Divide the message into second line, third line, fourth line
    separator_pos = other_lines.find('\n');
    std::string second_line(other_lines.substr(0, separator_pos));
    std::string third_fourth_line(other_lines.substr(separator_pos));
    separator_pos = third_fourth_line.find('\n');
    std::string third_line(third_fourth_line.substr(0, separator_pos));
    std::string fourth_line(third_fourth_line.substr(separator_pos));

    int adapter_id = std::stoi(second_line);
    uint32_t final_seq_no_control = std::stoul(third_line);
    uint32_t final_seq_no_data = std::stoul(fourth_line);

    this->on_receive_disconnect_message(adapter_id, final_seq_no_control,
                                        final_seq_no_data);
  }
  case CMCode::kCMCodePriv: {
    // Priv type
    this->on_receive_priv_message(other_lines);
    break;
  }
  default: {
    LOG_ERR("Unknown Control Message Code(%d)!\n%s", control_message_code,
            message.c_str());
    break;
  }
  }

  free(message_buffer);
  return true;
}

void ControlMessageReceiver::on_receive_normal_message(int control_message_code,
                                                       int adapter_id) {
  NetworkSwitcher *ns = NetworkSwitcher::singleton();
  switch (control_message_code) {
  case CMCode::kCMCodeConnect: {
    LOG_VERB("Receive(Control Msg): Request(Connect %d)", adapter_id);
    ns->connect_adapter_by_peer(adapter_id);
    break;
  }
  case CMCode::kCMCodeSleep: {
    LOG_VERB("Receive(Control Msg): Request(Sleep %d)", adapter_id);
    ns->sleep_adapter_by_peer(adapter_id);
    break;
  }
  case CMCode::kCMCodeWakeUp: {
    LOG_VERB("Receive(Control Msg): Request(WakeUp %d)", adapter_id);
    ns->wake_up_adapter_by_peer(adapter_id);
    break;
  }
  case CMCode::kCMCodeDisconnectAck: {
    LOG_VERB("Receive(Control Msg): Request(DisconnectAck %d)", adapter_id);
    ServerAdapter *disconnect_adapter =
        Core::singleton()->find_adapter_by_id((int)adapter_id);
    if (disconnect_adapter == NULL) {
      LOG_WARN("Cannot find adapter %d", (int)adapter_id);
    } else {
      disconnect_adapter->peer_knows_disconnecting_on_purpose();
    }
    break;
  }
  default: {
    // Never reach here
    break;
  }
  }
}

void ControlMessageReceiver::on_receive_disconnect_message(
    int adapter_id, uint32_t final_seq_no_control, uint32_t final_seq_no_data) {
  LOG_VERB("Receive(Control Msg): Request(Disconnect %d / "
           "final_seq_no_control=%lu / final_seq_no_data=%lu)",
           adapter_id, final_seq_no_control, final_seq_no_data);
  NetworkSwitcher *ns = NetworkSwitcher::singleton();
  ns->disconnect_adapter_by_peer(adapter_id, final_seq_no_control,
                                 final_seq_no_data);
}

void ControlMessageReceiver::on_receive_priv_message(std::string contents) {
  // Find separator location (between second line and other lines)
  std::size_t separator_pos = contents.find('\n');

  // Divide the message into second line & other lines
  std::string second_line(contents.substr(0, separator_pos));
  std::string priv_message(contents.substr(separator_pos));

  int priv_type = std::stoi(second_line);

  // Notify the priv message
  for (std::vector<ControlPrivMessageListener *>::iterator it =
           this->mPrivMessageListeners.begin();
       it != this->mPrivMessageListeners.end(); it++) {
    ControlPrivMessageListener *listener = *it;
    if (listener != NULL) {
      listener->on_receive_control_priv_message(priv_type, priv_message);
    }
  }
}